jekyll/_cci2/optimization-cookbook.md

```diff
@@ -0,0 +1,210 @@
+
+If a job fetches data at any point, it is likely that you can make use of caching. A common example is the use of a package/dependency manager. If your project uses Yarn, Bundler, or Pip, for example, the dependencies downloaded during a job can be cached for later use rather than being re-downloaded on every build. The example below shows how you can use caching for a package manager.
+
+{% raw %}
+```yaml
+version: 2
+jobs:
+  build:
+    steps: # a collection of executable commands making up the 'build' job
+      - checkout # pulls source code to the working directory
+      - restore_cache: # **restores saved dependency cache if the Branch key template or requirements.txt files have not changed since the previous run**
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+      - run: # install and activate virtual environment with pip
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            pip install -r requirements.txt
+      - save_cache: # ** special step to save dependency cache **
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+          paths:
+            - "venv"
+```
+{% endraw %}
+
+Notice in the above example that you can use a `checksum` in the cache key. This is used to calculate when a specific dependency-management file (such as a `package.json` or `requirements.txt` in this case) changes so the cache will be updated accordingly. Also note that the `restore_cache` example uses interpolation to put dynamic values into the cache-key, allowing more control in what exactly constitutes the need to update a cache.
+
+**Note:** Before adding any caching steps to your workflow, verify the dependencies installation step succeeds. Caching a failed dependency step will require you to change the cache key in order to avoid failed builds due to a bad cache. 
+
+Because caching is a such a critical aspect of optimizing builds and workflows, you should first familiarize yourself with the following page that describes caching and how various strategies can be optimized:
+
+- Caching - https://circleci.com/docs/2.0/caching/
+
+### Caching Example
+
+<Example needed here>
+
+## Testing Optimization
+
+When running tests on the CircleCI platform, one of the primary considerations you will want to make is how you can optimize the testing process to minimize credit usage and improve overall testing performance and results. Testing can sometimes be a time and performance-intensive process; therefore, the ability to reduce testing time can be a significant boost to your organizational goals.
+
+There are many different test suites and approaches you can use when testing on the CircleCI platform. Although CircleCI is test suite agnostic, the example below illustrates an example of how you could optimize testing using Django and the CircleCI platform.
+
+### Testing Optmization on the CircleCI Platform For a Python Django Project
+
```

Comment:
This section uses significant portions of my [blog post](https://www.brautaset.org/articles/2019/speed-up-circleci.html) without attribution. I would appreciate if you could work "adapted (with permission) from [this blog post](https://www.brautaset.org/articles/2019/speed-up-circleci.html)" into the text.

Reply:
@stig . Absolutely will make a note of this in the document so there is proper attribution. Thanks for the help on this :)